components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/TomcatCookieAuthenticationHandler.java

```diff
@@ -0,0 +1,128 @@
+import org.wso2.carbon.identity.auth.service.AuthenticationStatus;
+import org.wso2.carbon.identity.auth.service.handler.AuthenticationHandler;
+import org.wso2.carbon.identity.core.bean.context.MessageContext;
+import org.wso2.carbon.utils.ServerConstants;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import static org.apache.commons.lang.StringUtils.isNotBlank;
+import static org.wso2.carbon.identity.auth.service.util.Constants.COOKIE_AUTH_HEADER;
+import static org.wso2.carbon.identity.auth.service.util.Constants.JSESSIONID;
+
+/**
+ * This handler is used to authenticate the rest APIs based on the set-cookie obtained from the AuthenticationAdmin
+ * Service.
+ */
+public class TomcatCookieAuthenticationHandler extends AuthenticationHandler {
+
+    private static final Log log = LogFactory.getLog(TomcatCookieAuthenticationHandler.class);
+
+    @Override
+    public String getName() {
+
+        return "TomcatCookieAuthentication";
+    }
+
+    @Override
+    public int getPriority(MessageContext messageContext) {
+
+        int priority = super.getPriority(messageContext);
+        if (priority != -1) {
+            return priority;
+        }
+        return 500;
+    }
+
+    @Override
+    public boolean canHandle(MessageContext messageContext) {
+
+        if (messageContext instanceof AuthenticationContext) {
+            AuthenticationContext authenticationContext = (AuthenticationContext) messageContext;
+            if (authenticationContext.getAuthenticationRequest() != null) {
+                String cookie = authenticationContext.getAuthenticationRequest().
+                        getHeader(COOKIE_AUTH_HEADER);
+                return StringUtils.isNotEmpty(cookie) && cookie.startsWith(JSESSIONID);
+            }
+        }
+        return false;
+    }
+
+    @Override
+    protected AuthenticationResult doAuthenticate(MessageContext messageContext) {
+
+        AuthenticationResult authenticationResult = new AuthenticationResult(AuthenticationStatus.FAILED);
+        AuthenticationContext authenticationContext = (AuthenticationContext) messageContext;
+
+        if (servletRequestExists(authenticationContext)) {
+            Request request = (Request) authenticationContext.getAuthenticationRequest().getAttribute(HTTPConstants
+                    .MC_HTTP_SERVLETREQUEST);
+            if (isLoggedInUserExists(request)) {
+                String userName = (String) request.getSession().getAttribute(ServerConstants.USER_LOGGED_IN);
+                if (isNotBlank(userName)) {
+                    String tenantDomain = MultitenantUtils.getTenantDomain(userName);
+                    User user = buildUser(userName, tenantDomain);
+                    authenticationContext.setUser(user);
+                    authenticationResult.setAuthenticationStatus(AuthenticationStatus.SUCCESS);
+                    if (log.isDebugEnabled()) {
+                        log.debug("Tomcat Cookie Authentication success.");
```

Comment:
can we add more context info to the log.

Reply:
Authenticated user will be logged in other places